linter/lints/cpcps/lint_crl_has_idp.go

```diff
@@ -89,107 +80,108 @@ func (l *crlHasIDP) Execute(c *x509.RevocationList) *lint.LintResult {
-
-	if idpv.PeekASN1Tag(distributionPointTag) || idpv.PeekASN1Tag(onlyContainsUserCertsTag) {
-		// If either the distributionPoint [0] field or the
-		// onlyContainsUserCerts [1] field is present, assume we're dealing with
-		// a CRL containing Subscriber Certs.
-
-		var dpName cryptobyte.String
-		if !idpv.ReadASN1(&dpName, distributionPointTag) {
-			return &lint.LintResult{
-				Status:  lint.Warn,
-				Details: "Failed to read IDP distributionPoint",
-			}
-		}
-		if !dpName.ReadASN1(&dpName, distributionPointFullNameTag) {
-			return &lint.LintResult{
-				Status:  lint.Warn,
-				Details: "Failed to read IDP distributionPoint fullName",
-			}
+	if !idpv.ReadOptionalASN1(&dpName, &distributionPointExists, distributionPointTag) {
+		return &lint.LintResult{
+			Status:  lint.Warn,
+			Details: "Failed to read IssuingDistributionPoint distributionPoint",
 		}
+	}
 
-		var uriBytes []byte
-		if !dpName.ReadASN1Bytes(&uriBytes, distributionPointURITag) {
-			return &lint.LintResult{
-				Status:  lint.Warn,
-				Details: "Failed to read IDP URI",
-			}
+	idp := lints.NewIssuingDistributionPoint()
+	onlyContainsUserCertsTag := cryptobyte_asn1.Tag(1).ContextSpecific()
+	if !lints.ReadOptionalASN1BooleanWithTag(&idpv, &idp.OnlyContainsUserCerts, onlyContainsUserCertsTag, false) {
+		return &lint.LintResult{
+			Status:  lint.Error,
+			Details: "Failed to read IssuingDistributionPoint onlyContainsUserCerts",
 		}
+	}
 
-		uri, err := url.Parse(string(uriBytes))
-		if err != nil {
-			return &lint.LintResult{
-				Status:  lint.Error,
-				Details: "Failed to parse IDP URI",
-			}
-		}
-		if uri.Scheme != "http" {
-			return &lint.LintResult{
-				Status:  lint.Error,
-				Details: "IDP URI MUST use http scheme",
-			}
+	onlyContainsCACertsTag := cryptobyte_asn1.Tag(2).ContextSpecific()
+	if !lints.ReadOptionalASN1BooleanWithTag(&idpv, &idp.OnlyContainsCACerts, onlyContainsCACertsTag, false) {
+		return &lint.LintResult{
+			Status:  lint.Error,
+			Details: "Failed to read IssuingDistributionPoint onlyContainsCACerts",
 		}
-		if !dpName.Empty() {
-			return &lint.LintResult{
-				Status:  lint.Warn,
-				Details: "IDP should contain only one distributionPoint",
-			}
+	}
+
+	if !idpv.Empty() {
+		return &lint.LintResult{
+			Status:  lint.Error,
+			Details: "Unexpected IssuingDistributionPoint fields were found",
 		}
+	}
 
-		tagPresent := false
-		ocucPresent := false
-		ok := lints.ReadOptionalASN1BooleanWithTag(&idpv, &tagPresent, &ocucPresent, onlyContainsUserCertsTag, false)
-		if !ok || !tagPresent {
+	if idp.OnlyContainsUserCerts {
+		if idp.OnlyContainsCACerts {
 			return &lint.LintResult{
 				Status:  lint.Error,
-				Details: "Failed to read IDP onlyContainsUserCerts",
+				Details: "IssuingDistributionPoint should not have both onlyContainsUserCerts: TRUE and onlyContainsCACerts: TRUE",
 			}
 		}
-		if !ocucPresent {
+		lintErrors := parseSingleDistributionPointName(dpName, idp)
```

Comment:
This function has now mostly been split into parsing (up top) and semantic validation (down below). The call to `parseSingleDistributionPointName` is one exception. Can you move it up above where the parsing happens? The idea is that immediately after parsing out `dpName` from the input, you would pass `dpName` to `parseSingleDistributionPointName` to parse its contents.

Reply:
For some reason I didn't think to do that myself even though I recognized that having `parseSingleDistributionPointName` at that spot was odd. Thanks!